src/config.rs

```diff
@@ -2671,6 +2671,11 @@ impl TiKvConfig {
         {
             return Err("default rocksdb doesn't exist, but raftdb or raft engine exists".into());
         }
+        if RocksEngine::exists(&self.raft_store.raftdb_path)
```

Comment:
What if TiKV crashes during upgrading data from rocksdb to raft engine?

Reply:
I'll add a deferred cleanup. But in case the directory is still not cleaned up due to filesystem error, user has to manually do it.

Reply:
How about using a temporary directory and rename to do atomic upgrade?

Reply:
I think your solution is safer, but it involves some extra work to reopen the db. Plus it won't prevent coexist if filesystem does err (rename of target directory succeeds, rename of source directory fails).

Reply:
I think the process should be:
1. Upgraded to tempdir (you can even used a fixed name), let's refer as raft0;
2. Rename rocksdb directory to a temp directory, let's refer it as rocks1 -> rocks0;
3. Rename tempdir to raft engine directory, let's refer it as raft0 -> raft1;
4. Delete rocksdb directory.

So no matter which steps it crashes, it can still resume automatically. Note following situations are impossible:
1. raft0 and raft1 coexist;
2. rocks0 and rocks1 coexist;
3. raft1 and rocks1 coexist.

So if raft0 and rocks0 coexist, just resume from step 3. If rocks1 exists, delete raft0 if exists and resume from step 1. If raft1 exists, delete rocks0 if exists, and finish upgrading.

Reply:
Nice, but if we allow a weird intermediate state to exist, I can do better without reopening DB:
[A] -> [A.0] -> [A.0, B] -> [B] (A.0 takes priority over any directory)

Maybe I'll try it later.

Reply:
I have to say I don't understand what you mean.

Reply:
A/B refers to your raft1/rocks1, A.0 refers to your raft0/rocks0.

Reply:
I don't think there is an atomic way to delete `A.0`. It's totally possible that only some files in `A.0` are deleted before TiKV crashes.

Reply:
Oh, just add another rename. The old code already does it this way.

Reply:
On a second thought, I think this introduces a lot of complexities, especially to the config check and engine init part. I'll stick to the current solution.

Reply:
> introduces a lot of complexities, especially to the config check

Why? What it needs to do is just guarantee the three asserts listed above.

> and engine init part

It just needs to follow the migration order mentioned above.

Reply:
We might have different understanding for "complex". I consider gluing config check and engine dumping with a shared state machine a serious complexity. And I prefer not to do it given the current solution already prevents data corruption.

Reply:
So we know it's a possible state, and we know the state is fine, and the correct way to fix the problem is just delete the directory. However, somehow we choose to have someone to be called up to handle it instead of adding about tens of lines? 🤣

Reply:
I understand this sentiment. But the possibility of failing to rm a directory (while succeed in opening a DB) is pretty low. And we give clear instruction to the user (rm a specific directory manually). Meanwhile a decent implementation (without looking like a patchwork) of the novel solution we discussed would probably takes hundreds of lines.

Reply:
sigh.. Just noticed we use `process::abort` in panic hook, so that defer block won't be called. Maybe I still have to do it the hard way.